ios/Exponent/Kernel/Services/EXUserNotificationManager.m

```diff
@@ -0,0 +1,91 @@
+  dispatch_once(&once, ^{
+    if (!theManager) {
+      theManager = [EXUserNotificationManager new];
+    }
+  });
+  return theManager;
+}
+
+- (void)userNotificationCenter:(UNUserNotificationCenter *)center
+didReceiveNotificationResponse:(UNNotificationResponse *)response
+         withCompletionHandler:(void (^)(void))completionHandler
+{
+  BOOL isFromBackground = !([UIApplication sharedApplication].applicationState == UIApplicationStateActive);
+  NSDictionary *payload = response.notification.request.content.userInfo;
+  if (payload) {
+    NSDictionary *body = payload[@"body"];
+    NSString *experienceId = payload[@"experienceId"];
+    NSString *userText = @"";
+    NSString *actionId = @"DEFAULT_ACTION";
+    
+    if ([response.actionIdentifier isEqualToString:UNNotificationDismissActionIdentifier]) {
+      actionId = @"DISMISS_ACTION";
+    } else if (![response.actionIdentifier isEqualToString:UNNotificationDefaultActionIdentifier]) {
+      actionId = response.actionIdentifier;
+      if (![EXEnvironment sharedEnvironment].isDetached) {
+        actionId = [actionId componentsSeparatedByString:@":"][1];
+      }
+    }
+    
+    if ([response isKindOfClass:[UNTextInputNotificationResponse class]]) {
+      userText = ((UNTextInputNotificationResponse *) response).userText;
+    }
+    
+    BOOL isRemote = [response.notification.request.trigger isKindOfClass:[UNPushNotificationTrigger class]];
+    if (body && experienceId) {
+      [[EXKernel sharedInstance] sendNotification:body
+                               toExperienceWithId:experienceId
+                                   fromBackground:isFromBackground
+                                         isRemote:isRemote
+                                         actionId:actionId
+                                         userText:userText];
+    }
+  }
+  completionHandler();
+}
+
+- (void)userNotificationCenter:(UNUserNotificationCenter *)center
+       willPresentNotification:(UNNotification *)notification
+         withCompletionHandler:(void (^)(UNNotificationPresentationOptions options))completionHandler
+{
+  NSDictionary *payload = notification.request.content.userInfo;
+  if (payload) {
+    NSDictionary *body = payload[@"body"];
+    NSString *experienceId = payload[@"experienceId"];
+    NSString *userText = @"";
+    
+    BOOL isRemote = [notification.request.trigger isKindOfClass:[UNPushNotificationTrigger class]];
+    if (body && experienceId) {
+      [[EXKernel sharedInstance] sendNotification:body
+                               toExperienceWithId:experienceId
+                                   fromBackground:NO
+                                         isRemote:isRemote
+                                         actionId:@"WILL_PRESENT_ACTION"
+                                         userText:userText];
+    }
+  }
+  completionHandler(UNNotificationPresentationOptionAlert + UNNotificationPresentationOptionSound);
+}
+
+- (void)setNotificationDelegate
```

Comment:
Name of this method isn't the best description of its content. Maybe we should just set the delegate in the initializer? I suppose this is how it's done in EXAnalytics — https://github.com/expo/expo/pull/2316/files#diff-0c6b818bdfc75ebed906c186926f43c4R102

Reply:
bump